File: 0226-invert-binary-tree/0226-invert-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */

//  1️⃣ Recursive DFS (Preorder - Most Common & Optimal)
// Time Complexity: O(N) → Visits every node once.

// Space Complexity: O(H) → Recursive call stack (H = height of tree).
// var invertTree = function(root) {
//     if (root === null) return null; // Base case

//     // Swap left and right subtrees
//     let temp = root.left;
//     root.left = root.right;
//     root.right = temp;

//     // Recursively invert left and right subtrees
//     invertTree(root.left);
//     invertTree(root.right);

//     return root;
// };

var invertTree = function(root) {
 if (!root) return null;
    let stack=[root];
    while(stack.length > 0){
        let node=stack.pop();

        [node.left, node.right]=[node.right, node.left];
          if(node.left !=null) stack.push(node.left);
        if(node.right !=null) stack.push(node.right);
    }
    return root;
}

// 3️⃣ BFS (Using Queue - Level Order)
// Time Complexity: O(N) → Visits every node once.
// Space Complexity: O(W) → Stores W nodes (W = max width of tree).
// var invertTree = function(root) {
//  if (!root) return null;
//     let queue = [root];
//     while(queue.length > 0){
//         let node =queue.shift();
//        [node.left, node.right] = [node.right, node.left];

//         if(node.left !=null) queue.push(node.left);
//         if(node.right !=null) queue.push(node.right);
//     }
//     return root
// }

File: 0226-invert-binary-tree/0226-invert-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// TC : O(n), where n is the number of nodes in the binary tree.
// SC : In the worst case (skewed tree), the depth of recursion is O(n).
// In the best case (balanced tree), it's O(log n).
class Solution {
    public TreeNode invertTree1(TreeNode root) {
        if (root == null)
            return null;

        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;

        invertTree(root.left);
        invertTree(root.right);

        return root;
    }

    public TreeNode invertTree2(TreeNode root) {
        if (root == null)
            return null;
        Queue<TreeNode> qu = new LinkedList<TreeNode>();
        qu.offer(root);
        while (!qu.isEmpty()) {
            TreeNode curr = qu.poll();
            TreeNode temp = curr.left;
            curr.left = curr.right;
            curr.right = temp;

            if (curr.left != null)
                qu.offer(curr.left);
            if (curr.right != null)
                qu.offer(curr.right);
        }
        return root;
    }

     public TreeNode invertTree(TreeNode root) {
            if (root == null)
            return null;
        Stack<TreeNode> st = new Stack<TreeNode>();
        st.add(root);
        while(!st.isEmpty()){
            TreeNode curr=st.pop();
            TreeNode temp = curr.left;
            curr.left=curr.right;
            curr.right=temp;

            if(curr.left !=null) st.add(curr.left);
            if(curr.right !=null) st.add(curr.right);
        }
        return root;
     }
}